### ISSUE TITLE
Replace testNoAlloc with something more reliable

### ISSUE THREAD

**Issue by AndrejMitrovic** [bosagora/agora#699]
I got this random bug right now:

```
ocean.core.Test.TestException@source/agora/common/Serializer.d(740): 
Expression expected to not allocate but GC usage stats have changed from 
4158064 (used) / 1084816 (free) to 4158320 / 1084560
```

It triggers randomly, of course.

We had this problem before except it was unittests in `ocean` that were running in parallel to the agora tests. We stopped running the `ocean` unittests but now the call to `testNoAlloc()` was added to `Serializer.d` so the problem reappeared.

Why does `testNoAlloc` exist? Is it like an escape hatch when we want to verify a function doesn't allocate but we just can't use `@nogc`?

**Comment by Geod24** [comment]
> Why does `testNoAlloc` exist? Is it like an escape hatch when we want to verify a function doesn't allocate but we just can't use `@nogc`?

Yes. Consider a function that can allocate on its error path. Or a function that takes a delegate, which itself can't be `@nogc` because then that would restrict what the delegate can do, but is known to be `@nogc` by the caller.

So if I understand correctly, the problem is that tests run in parallel ? One thing I'd love to have is the ability to poison calls, so that anything that calls `gc_malloc` triggers an `abort`.

**Comment by AndrejMitrovic** [comment]
> So if I understand correctly, the problem is that tests run in parallel ?

Yep, and some tests naturally allocate.

Can we somehow hook into `gc_malloc`?

Alternatively maybe some of the D test frameworks could allow us to run some tests with a single thread, while others in parallel. Something like `@singlethreaded unittest { }` that runs after all the parallelized tests run would be neat.

**Comment by Geod24** [comment]
So https://github.com/bpfkorea/agora/issues/402

**Comment by Geod24** [comment]
`testNoAlloc` ever caused problems once, because of a lazy allocation.
It was fairly easy to track down and it has been a great help since (it caught a Phobos bug in `BitBlob`, for example).
I think we can close this issue as WONTFIX, since there are much bigger issues to tackle.
